<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

**Exercise 1**

In [24]:
list = []
for number in range(10):
    if number % 2 ==0:
        continue
        
    number **=2
 
    list.append(number)
print(list)

[1, 9, 25, 49, 81]


**Exercise 2**

In [22]:
[number**2 for number in range(10) if number %2 != 0]

[1, 9, 25, 49, 81]

**Exercise 3**

In [25]:
%%timeit 
[number**2 for number in range(10) if number %2 != 0]

902 ns ± 80.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [26]:
%%timeit
list = []
for number in range(10):
    if number % 2 ==0:
        continue
        
    number **=2
 
    list.append(number)

1.02 μs ± 53 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


The code using list comprehension in exericse 2 ran faster, at 902 nanoseconds per loop, while the code from exercise 1 took 1.02 microseconds per loop.

**Exercise 4**

In [37]:
substance = 100
years = 0
while substance >= 50:
    substance = (95/100)*substance
    years += 1
    
print(years,'years')

14 years


**Exercise 5**

In [91]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

counter = 0
for year, CO2_level in co2_data:
    if counter == len(co2_data) - 1:  #len(co2_data) - 1 because the list has 21 varaibles and below                             
        break                         #co2_data[counter +1] has a range till co2_data[20], stop loop at 
                                      #[counter + 1] = 20
    diff_in_co2 = (co2_data[counter + 1][1]) - (CO2_level)  #diff between the current year in the loop and the year after
    
    if  diff_in_co2 >= 3:
        print(f'{year}/'+str(co2_data[counter + 1][0])+':','+' + f'{diff_in_co2:.2f}'+'ppm')
                                      #made co2_data a str to join the varaibles together for aesthetic purposes..
            
    counter += 1
    
    

2015/2016: +3.41ppm
